File: pset6/credit/credit.py
import sys
user_input = input("Enter a credit card number: ")

if not len(user_input) in [13, 15, 16]:
    print("INVALID")
    sys.exit()


# implement Lunh Algorithm
x = int(user_input)
sum1 = 0
sum2 = 0
totalSum = 0
while True:
    # get every other digit starting from last
    rem1 = x % 10
    x = int(x / 10)
    sum1 = sum1 + rem1

    # get second to last digit
    rem2 = x % 10
    x = int(x / 10)

    # double digits and add to sum
    rem2 = rem2 * 2
    dig1 = rem2 % 10
    dig2 = int(rem2 / 10)
    sum2 = sum2 + dig1 + dig2
    if x <= 0:
        break

totalSum = sum1 + sum2

# check if last digit of sum equals 0
if totalSum % 10 != 0:
    print("INVALID")
    sys.exit()


# check for starting digits
first_two = int(user_input[0:2])
first_digits = int(user_input)

if first_two == 37 or first_two == 34:
    print("AMEX")
elif first_two > 50 and first_two <= 55:
    print("MASTERCARD")
elif first_two // 10 == 4:
    print("VISA")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    //ask user for card number
    long num = get_long("Enter a credit card number ");
    //count the digits
    int dig = 0;
    long card = num;
    while (card > 0)
    {
        card = card / 10;
        dig++;
    }

    if (dig != 13 && dig != 15 && dig != 16) //check the lenght of number to see if it could be a credit card
    {
        printf("INVALID\n");
        return 0;
    }

    //calculate checksum
    int sum1 = 0;
    int sum2 = 0;
    long x = num;
    int totalSum = 0;
    int rem1;
    int rem2;
    int dig1;
    int dig2;

    do
    {
        //get every other digit starting from the last digit
        rem1 = x % 10;
        x = x / 10;
        sum1 = sum1 + rem1;

        //get second to last digit
        rem2 = x % 10;
        x = x / 10;

        //double digits and add to sum
        rem2 = rem2 * 2;
        dig1 = rem2 % 10;
        dig2 = rem2 / 10;
        sum2 = sum2 + dig1 + dig2;
    }
    while (x > 0);
    totalSum = sum1 + sum2;
    //check if last digit of sum equals 0
    if (totalSum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    //check for starting digits
    long firstDigits = num;
    do
    {
        firstDigits /= 10;
    }
    while (firstDigits > 100);
    if (firstDigits == 37 || firstDigits == 34)
    {
        printf("AMEX\n");
    }

    else if (firstDigits > 50 && firstDigits <= 55)
    {
        printf("MASTERCARD\n");
    }

    else if ((firstDigits / 10) == 4)
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}